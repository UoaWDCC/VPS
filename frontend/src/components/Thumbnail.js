import React from "react";

function Thumbnail({
  url = "https://canvas.ac.nz/",
  iframeWidth = 2000,
  iframeHeight = 1500,
  width = 200,
  height = 150,
  interactive = false,
}) {
  const xscale = width / iframeWidth;
  const yscale = height / iframeHeight;
  const iframe = {
    transform: `scaleX(${xscale}) scaleY(${yscale})`,
    transformOrigin: "0 0",
    width: `${iframeWidth}px`,
    height: `${iframeHeight}px`,
    border: "0px",
  };
  const iframeShade = {
    position: "absolute",
    width: "100%",
    height: `${height}px`,
    maxWidth: `${width}px`,
    zIndex: "10",
  };

  const iframeWrapper = {
    maxWidth: `${width}px`,
    maxHeight: `${height}px`,
    backgroundColor: "white",
    filter: "drop-shadow(1px 1px 2px grey)",
  };

  const outerWrapper = {
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    width: "100%",
    height: "100%",
  };

  return (
    <div style={outerWrapper}>
      <div style={iframeWrapper}>
        {!interactive && <div style={iframeShade}> </div>}
        <iframe
          tabIndex="-1"
          style={iframe}
          title="asdf"
          src={url}
          sandbox="allow-scripts allow-same-origin"
          scrolling="no"
        />
      </div>
    </div>
  );
}

export default Thumbnail;
