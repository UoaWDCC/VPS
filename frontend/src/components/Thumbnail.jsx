import { useState } from "react";
import Skeleton from "react-loading-skeleton";

/**
 * Component used for thumbnails.
 *
 * @component
 * @example
 * const url = "https://canvas.ac.nz/"
 * const interactive = false
 * function onLoad() {
 *   console.log("Loading...")
 * }
 * return (
 *   <Thumbnail url={url} interactive={interactive} onLoad={onLoad} />
 * )
 */
function Thumbnail({
  url = "https://canvas.ac.nz/",
  iframeWidth = 1920,
  iframeHeight = 1080,
  width = 266,
  height = 150,
  interactive = false,
  style: customStyle,
  onLoad,
}) {
  const [loading, setLoading] = useState(true);

  const xscale = width / iframeWidth;
  const yscale = height / iframeHeight;
  const iframe = {
    transform: `scaleX(${xscale}) scaleY(${yscale})`,
    transformOrigin: "0 0",
    width: `${iframeWidth}px`,
    height: `${iframeHeight}px`,
    border: "0px",
  };

  const iframeShade = {
    position: "absolute",
    width: "100%",
    height: `${height}px`,
    maxWidth: `${width}px`,
    zIndex: "10",
  };

  const iframeWrapper = {
    maxWidth: `${width}px`,
    maxHeight: `${height}px`,
    backgroundColor: "white",
    filter: "drop-shadow(1px 1px 2px grey)",
  };

  const outerWrapper = {
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    width: "100%",
    height: "100%",
  };

  /** Function which executes while the thumbnail is loading. */
  const combinedOnLoad = () => {
    if (onLoad) {
      onLoad();
    }
    setLoading(false);
  };

  return (
    <div
      style={{ ...outerWrapper, ...customStyle }}
      data-testid="div-outer-wrapper"
    >
      <div style={iframeWrapper} data-testid="div-iframe-wrapper">
        {!interactive && <div style={iframeShade}> </div>}
        {loading && (
          <Skeleton
            style={{ position: "absolute", zIndex: "15" }}
            width={width}
            height={height}
            duration={0.8}
          />
        )}
        <iframe
          tabIndex="-1"
          style={iframe}
          title="webpage-thumbnail"
          src={url}
          sandbox="allow-scripts allow-same-origin"
          scrolling="no"
          onLoad={combinedOnLoad}
        />
      </div>
    </div>
  );
}
// TODO write proptypes
export default Thumbnail;
