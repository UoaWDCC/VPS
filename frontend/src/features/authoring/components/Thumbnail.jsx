import Box from "../elements/Box";
import Ellipse from "../elements/Ellipse";
import Image from "../elements/Image";
import Line from "../elements/Line";
import Speech from "../elements/Speech";
import TextBox from "../elements/TextBox";
import { buildVisualComponents, buildVisualScene } from "../pipeline";

const componentMap = {
  textbox: TextBox,
  speech: Speech,
  ellipse: Ellipse,
  box: Box,
  image: Image,
  line: Line,
};

function resolve(component) {
  const Fc = componentMap[component.type];
  if (Fc) return <Fc key={component.id} {...component} />;
  return null;
}

const Thumbnail = ({ components }) => {
  const visual = buildVisualComponents(components);
  const visualComponents = visual
    .sort((a, b) => a.zIndex - b.zIndex)
    .map(resolve);

  return (
    <svg viewBox="0 0 1920 1080">
      <rect x="0" y="0" width="1920" height="1080" fill="white" />
      {visualComponents}
    </svg>
  );
};

export default Thumbnail;
