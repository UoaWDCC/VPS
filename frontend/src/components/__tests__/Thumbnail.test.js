import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import Thumbnail from "../Thumbnail";

test("Thumbnail component snapshot test", () => {
  const component = renderer.create(
    <BrowserRouter>
      <Thumbnail />
    </BrowserRouter>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
