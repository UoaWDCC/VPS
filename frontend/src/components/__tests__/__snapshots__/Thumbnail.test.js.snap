// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Thumbnail component snapshot test 1`] = `
<div
  data-testid="div-outer-wrapper"
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "height": "100%",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <div
    data-testid="div-iframe-wrapper"
    style={
      Object {
        "backgroundColor": "white",
        "filter": "drop-shadow(1px 1px 2px grey)",
        "maxHeight": "150px",
        "maxWidth": "266px",
      }
    }
  >
    <div
      style={
        Object {
          "height": "150px",
          "maxWidth": "266px",
          "position": "absolute",
          "width": "100%",
          "zIndex": "10",
        }
      }
    >
       
    </div>
    <span>
      <span
        className="react-loading-skeleton css-bvijti-skeletonStyles-Skeleton"
        style={
          Object {
            "height": 150,
            "position": "absolute",
            "width": 266,
            "zIndex": "15",
          }
        }
      >
        â€Œ
      </span>
    </span>
    <iframe
      onLoad={[Function]}
      sandbox="allow-scripts allow-same-origin"
      scrolling="no"
      src="https://canvas.ac.nz/"
      style={
        Object {
          "border": "0px",
          "height": "1080px",
          "transform": "scaleX(0.13854166666666667) scaleY(0.1388888888888889)",
          "transformOrigin": "0 0",
          "width": "1920px",
        }
      }
      tabIndex="-1"
      title="webpage-thumbnail"
    />
  </div>
</div>
`;
